pmd-core/src/main/java/net/sourceforge/pmd/lang/metrics/internal/AbstractLanguageMetricsProvider.java

```diff
@@ -0,0 +1,90 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.metrics.internal;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import net.sourceforge.pmd.lang.ast.Node;
+import net.sourceforge.pmd.lang.ast.QualifiableNode;
+import net.sourceforge.pmd.lang.metrics.LanguageMetricsProvider;
+import net.sourceforge.pmd.lang.metrics.MetricKey;
+import net.sourceforge.pmd.lang.metrics.MetricOptions;
+import net.sourceforge.pmd.lang.metrics.MetricsComputer;
+import net.sourceforge.pmd.lang.metrics.ResultOption;
+
+
+/**
+ * Base implementation for {@link LanguageMetricsProvider}.
+ *
+ * @author Clément Fournier
+ * @since 7.0.0
+ */
+public abstract class AbstractLanguageMetricsProvider<T extends QualifiableNode, O extends QualifiableNode> implements LanguageMetricsProvider<T, O> {
```

Comment:
Basically similar question like for JavaCCTokenizer, but opposite case :smile:
Shouldn't this be *not* internal? It's used to implement Metrics support in the language, so it's nice to have this available for subclassing.
Since the interface is `@Experimental`, we should mark this abstract class as experimental too, then.

Reply:
I guess the bigger question is whether implementing language supports should be a published API or not. The interfaces to use languages are definitely published API. But I'd say if we're the only ones to implement languages, which is likely, then the abstract classes and implementation classes should not be published API and only the interfaces should be. I don't see the point in publicly supporting an API that's probably next to unused outside of PMD. Doing so ties our hands more than necessary, and forces us to expose many implementation details that are irrelevant to users of our other APIs. Wdyt?